llms/mlx_lm/server.py

```diff
@@ -5,43 +5,38 @@
 import time
 import uuid
 import warnings
-from collections import namedtuple
 from http.server import BaseHTTPRequestHandler, HTTPServer
-from typing import Callable, List, Optional
+from typing import Callable, List, Literal, NamedTuple, Optional, Union
 
 import mlx.core as mx
 import mlx.nn as nn
-import numpy as np
 from transformers import PreTrainedTokenizer
 
 from .utils import generate_step, load
 
-_model: Optional[nn.Module] = None
-_tokenizer: Optional[PreTrainedTokenizer] = None
+MODEL: nn.Module
+TOKENIZER: PreTrainedTokenizer
 
+SYSTEM_FINGERPRINT: str = f"fp_{uuid.uuid4()}"
 
-def load_model(model_path: str, adapter_file: Optional[str] = None):
-    global _model
-    global _tokenizer
-    _model, _tokenizer = load(model_path, adapter_file=adapter_file)
-
-
-StopCondition = namedtuple("StopCondition", ["stop_met", "trim_length"])
+class StopCondition(NamedTuple):
+    stop_met: bool
+    trim_length: int
 
 
 def stopping_criteria(
     tokens: List[int],
-    stop_id_sequences: List[np.ndarray],
-    eos_token_id: int,
+    stop_id_sequences: List[List[int]],
```

Comment:
Isn't the stop_id_sequences a `List[np.ndarray]`? what's the reason you want to change it to `List[List[int]]` ?

Reply:
We used to call tokenizer decode and encode with 
```python
return_tensors="np"
```
However I did some testing, and not converting it to np made it 20% faster. The np.array_equal also checks List[int] against a np.ndarray, which is just as fast as checking 2 python lists for equality.

The performance difference is almost negligible, but we get to drop the numpy import entirely with this change, so we do gain something out of it.

Reply:
I generated random length random character combinations and called tokenizer.encode once with `return_tensors="np"` and once without
```python
import random
import string
import timeit

import numpy as np
import mlx.core as mx

from utils import load

MODEL, TOKENIZER = load("mlx-community/quantized-gemma-2b")

# Generate data to encode
words = [''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(random.randint(1, 10))) for _ in range(100)]

comparison = [141, 1321, 231, 31]

def test_tokenizer():
    tokens = TOKENIZER.encode(random.choice(words), add_special_tokens=False)
    if tokens == comparison:
        return



def test_np_tokenizer():
    tokens = TOKENIZER.encode(random.choice(words), return_tensors="np", add_special_tokens=False)
    if np.array_equal(tokens, comparison):
        pass


print("With np:    ", timeit.timeit(test_np_tokenizer, number=1000000))
print("Without np: ", timeit.timeit(test_tokenizer, number=1000000))
```
My results were:
```
With np:     12.122312665997015
Without np:  9.860362332998193
```